gmso/core/views.py

```diff
@@ -0,0 +1,180 @@
+            visited[type(potential_type)].add(potential_id)
+
+            yield potential_type
+
+
+class PotentialFilters:
+    UNIQUE_NAME_CLASS = "unique_name_class"
+    UNIQUE_EXPRESSION = "unique_expression"
+    UNIQUE_PARAMETERS = "unique_parameters"
+
+
+potential_identifiers = {
+    PotentialFilters.UNIQUE_NAME_CLASS: get_name_or_class,
+    PotentialFilters.UNIQUE_EXPRESSION: lambda p: str(p.expression),
+    PotentialFilters.UNIQUE_PARAMETERS: get_parameters,
+}
+
+
+class TopologyPotentialView:
+    """A potential view based on different filters for a topology's potentials.
+
+    Parameters
+    ----------
+    iterator: typing.Iterator, required=True
+        An iterator of either topology sites or connections from which to extract potentials from
+
+    Parameters
+    ----------
+    filter_by: str or function, default=None
+        If provided, filter the collected potentials by some unique identifier
+        of a potential.
+
+    Notes
+    -----
+    If `filter_by` is provided and is a custom function, the collected potentials are
+    filtered on the basis of the return value of the `filter_by` function. A single potential
+    is passed to the filter_by function and it should return an identifier (should be hashable)
+    for that potential thus describing its uniqueness in the context of which the filter is being
+    used. Some simple examples are given below.
+
+    Examples
+    --------
+    To use a TopologyPotentialView, a Topology must have few sites with AtomTypes or BondTypes.
+
+    >>> from gmso.core.topology import Topology
+    >>> from gmso.core.atom import Atom
+    >>> from gmso.core.atom_type import AtomType
+    >>> top = Topology(name="ViewTopology")
+    >>> sites = [Atom(name=f"Atom_{j}") for j in range(10)]
+    >>> atom_type1 = AtomType(name='atom_type1')
+    >>> atom_type2 = AtomType(name='atom_type2')
+    >>> for site in sites:
+    ...     site.atom_type = atom_type1 if int(site.name[-1]) % 2 == 0 else atom_type2
+    ...     top.add_site(site)
+    >>> top.update_topology()
+    >>> for atom_type in top.atom_types:
+    ...     print(atom_type.name)
+    atom_type1
+    atom_type2
+    >>> top.get_index(atom_type2)
+    1
+
+    Notes
+    -----
+    The implementation of this class is inspired from networkx.classes.reportviews.NodeView by extending
+    the idea of a view with filteration capabilities. See the source for NodeView for further details
+
+    https://github.com/networkx/networkx/blob/12c1a00cd116701a763f7c57c230b8739d2ed085/networkx/classes/reportviews.py#L115-L279
+    """
+
+    def __init__(self, iterator, filter_by=None):
+        self.iterator = iterator
+        self.filter_by = filter_by
+
+    def __iter__(self):
+        yield from self.yield_view()
+
+    def index(self, item):
+        for j, potential in enumerate(self.yield_view()):
+            if potential is item:
+                return j
+
+    def _collect_potentials(self):
```

Comment:
I think the point was to use the filters to get the number of unique potentials. But if you just looked at len(gmso.Topology.atom_types()), you would get the number of potentials with no filtering applies. See the code below as an example.

```python
import mbuild as mb
from foyer import Forcefield
from gmso.external import from_parmed
ff = Forcefield(name="oplsaa")
mol = mb.load("CCO", smiles=True)
filled_box = mb.packing.fill_box(mol, 1, density=0.8)
applied_top = ff.apply(filled_box)
applied_top = from_parmed(applied_top)
print(len(applied_top.atom_types))

5
```

But the number of atomtypes should be 9, one for each sites. This is akin to repeats=False essentially.

Reply:
I think this can be achieved by `applied_top.atom_types(filter_by=PotentialFilters.UNIQUE_NAME_CLASS`
So I think we just need to clarify this in the docstring?

Reply:
Yeah. This has been an issue. However, a repeat behavior is akin the following. Essentially, should an atomtype that is associated with multiple sites, contribute to overall length
```python
atom_types_repeated = filter(lambda p: p is not None, map(lambda s: s.atom_type, top._sites))
```

Reply:
@bc118 do you have any thoughts here? I think the more common thing would be to just be worried about unique potentials. But currently, someone has to use the above workaround @umesh-timalsina posted to get all of the potentials. So if that's behavior users may need, it may be at least useful write that in. But I'm also okay with the way it's written currently, and if down the line someone wants that feature, it's easy to write a filter that will return that info.

Reply:
Many users will want only the unique force field parameters, and not care about the repeating ones (GOMC, LAMMPS...).  I'm sure someone will also want all of them though, at some point.  I am personally, more concerned about the former.

Reply:
@bc118. I definitely agree to and second your points. However, uniqueness here is engine/context and use-case specific. So, if we give it some thought, the idea behind these filters makes intuitive sense atleast to me. With this, we are essentially deferring the use of these filters to our writers or any end user using this package. With a reasonable default and some good builtin filters.

```py
class PotentialFilters:
    UNIQUE_NAME_CLASS = "unique_name_class"  # Potential Name/Class (Uniqueness as incorporated in our workflows)
    UNIQUE_EXPRESSION = "unique_expression"  # No Clear path but can be leveraged in more rigorous sympy checks for uniquness
    UNIQUE_PARAMETERS = "unique_parameters" # No Clear path but can be leveraged in more rigorous sympy checks for uniquness
    UNIQUE_ID = "unique_id"  # Default behavior (I.e. provide all the ones with unique object ids)
    REPEAT_DUPLICATES = "repeat_duplicates" # Last use case for duplcation

    @staticmethod
    def all():
        return set(
            f"{PotentialFilters.__name__}.{k}"
            for k, v in PotentialFilters.__dict__.items()
            if not k.startswith("__") and not callable(v)
        )


```